### PR TITLE
Resizable Image Previews

### LINKED ISSUE #782: Custom size node (for View Image)
I want to see the view image node much larger than the other nodes, but there is no way to stretch it right now.
Using the zoom all the time is inconvenient.

### PR DESCRIPTION
**Pull request by joeyballentine**
Self explanatory. Closes #782

Decisions I had to make:
- It now only encodes the largest image preview, up to 2048x resolution. Maybe in the future we could make this value customizable for performance or something.
- Outputs can now save their size in nodeState

image
image
image

### CODE REVIEW

File: src/common/common-types.ts [hunk 1]
```diff
@@ -278,6 +279,7 @@ export interface NodeData {
     readonly inputData: InputData;
     readonly groupState?: GroupState;
     readonly inputSize?: InputSize;
+    readonly outputSize?: OutputSize;
```
[the remark this answers is not in the document]
Comment (joeyballentine) [reply src/common/common-types.ts:282]: Do I have to do that in this PR? Afaik there aren't any nodes that have both a textarea input and an image preview output so the syncing is irrelevant right now, and that feels like a bigger overall change/refactor

[the remark this answers is not in the document]
Comment (RunDevelopment) [reply src/common/common-types.ts:282]: Alrighty.

**Comment by joeyballentine** [comment]
> Why did you remove that anyway?

Because the size of the image itself is now more than just a display thing. If you make the node full res and the UI switches it to the smaller image, it'll shrink the image being displayed. 

I'm not sure how to get around that besides always stretching the image to the bounds of the node, which I'm not sure could be made to retain aspect ratio. I'm open to suggestions

**Comment by joeyballentine** [comment]
I did figure out how to get around that btw. It works pretty well now

File: src/renderer/components/outputs/LargeImageOutput.tsx [hunk 2]
```diff
@@ -1,14 +1,19 @@
...
+import { Size } from '../../../common/common-types';
 import { GlobalVolatileContext } from '../../contexts/GlobalNodeState';
+import { SettingsContext } from '../../contexts/SettingsContext';
 import { useDevicePixelRatio } from '../../hooks/useDevicePixelRatio';
+import { useMemoArray } from '../../hooks/useMemo';
+import { DragHandleSVG } from '../CustomIcons';
 import { OutputProps } from './props';
 
-const IMAGE_PREVIEW_SIZE = 200;
+const IMAGE_PREVIEW_SIZE = 216;
```
Comment (RunDevelopment) [inline src/renderer/components/outputs/LargeImageOutput.tsx:16]: Why this specific number?

File: src/renderer/components/outputs/LargeImageOutput.tsx [hunk 3]
```diff
@@ -30,120 +35,251 @@ const pickImage = (previews: PreviewImage[], realSize: number) => {
...
+        useEffect(() => {
+            if (last?.previews) {
+                const biggestImage = last.previews[last.previews.length - 1];
+                const img = new window.Image();
+                img.src = biggestImage.url;
+                img.onload = () => {
+                    setMaxSize({
+                        width: img.width,
+                        height: img.height,
+                    });
```
Comment (RunDevelopment) [inline src/renderer/components/outputs/LargeImageOutput.tsx:72]: Q: Does this make sense? Why should users be allowed to make a comically large View Image node?

Practically speaking, it's also not very user-friendly. Image setting up a chain to your liking, including a large View Image node, and then chainner just auto-shrinks the node when you run the chain with a small image. After all, people that work on small pixel art sprites would probably like to see their sprite in big.

**Review by RunDevelopment — COMMENTED** [review]
Code looks good. Just a few questions.

File: src/renderer/components/outputs/LargeImageOutput.tsx (hunk 2 above)
Comment (joeyballentine) [reply src/renderer/components/outputs/LargeImageOutput.tsx:16]: I removed the 8px of padding from both sides, and added it to the image preview size. I noticed there was a bit you could expand the box without making the node bigger, which was a bit odd. So I just made it bigger by default instead.

File: src/renderer/components/outputs/LargeImageOutput.tsx (hunk 3 above)
Comment (joeyballentine) [reply src/renderer/components/outputs/LargeImageOutput.tsx:72]: So you're suggesting we scale the image to the size (maintaining aspect ratio) no matter what?

Comment (joeyballentine) [reply src/renderer/components/outputs/LargeImageOutput.tsx:72]: And keeping the max size something like 1024x instead?

Comment (RunDevelopment) [reply src/renderer/components/outputs/LargeImageOutput.tsx:72]: > So you're suggesting we scale the image to the size (maintaining aspect ratio) no matter what?

Yes, but I just realized that we didn't do that before. We just centered small images before. Just centering small images would be fine too.

But I was mainly saying that resizing the LargeImageOutput based on the preview image size is bad.

Comment (joeyballentine) [reply src/renderer/components/outputs/LargeImageOutput.tsx:72]: Gotcha. Well, I could certainly make it so it's always something like a 216-1024 range and then we just center anything smaller

Comment (RunDevelopment) [reply src/renderer/components/outputs/LargeImageOutput.tsx:72]: Sounds good.

**Comment by RunDevelopment** [comment]
Found some bugs:

1. Resizing when snap to grid is enabled: With 16px grid size, you cannot resize the image preview back to its smallest size.
2. Output sizes are not saved.

**Comment by joeyballentine** [comment]
> Resizing when snap to grid is enabled: With 16px grid size, you cannot resize the image preview back to its smallest size.

I don't think this is something I can really fix. There's actually a bunch of issues with snap to grid all around chaiNNer, like iterator child nodes popping out of the iterator container. IMO this is a non-issue and the real issue is we need to better support snap to grid somehow

**Review by RunDevelopment — APPROVED** [review]

**Pull request merged**